_posts/2024-05-27-A-deep-dive-into-faster-semantic-sparse-retrieval-in-OS-2.12.md

```diff
@@ -0,0 +1,144 @@
+
+## Neural sparse search basics
+
+For neural sparse search, the fundamental workflow is very similar to BM25. Let's compare neural sparse search to BM25 in terms of latency at ingestion time and query time. 
+
+### Ingestion time
+
+At ingestion time, neural sparse search processes text differently compared to BM25. The most significant difference is that neural sparse search uses neural models to transform the raw text into tokens and their corresponding weights. These token/weight pairs comprise a _sparse vector_. The neural model expands the tokens with synonyms and generates token weights based on their semantic importance. Next, similarly to BM25, Lucene creates a posing list for each token. The only difference in this process is how the fields are stored. Neural sparse search stores the fields in the Lucene [FeatureField](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/document/FeatureField.java), while BM25 stores a standard OpenSearch text field in the Lucene [Field](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/document/Field.java). The difference in storage and its impact on search latency are negligible. As we discuss in the following sections, in most cases, **the bottleneck for ingestion throughput is model inference**.
+
+### Query time
+
+At query time, the search process consists of two parts: building the sparse vector and searching the inverted index. Neural sparse search has two modes: _doc-only_ and _bi-encoder_:
+
+- For the **doc-only** mode, we use a tokenizer and a predefined lookup table to build the sparse vector. The tokenization takes less than 1 millisecond. In this mode, the query tokens are not expanded with synonyms so all tokens must occur in the original query text. 
+- For the **bi-encoder** mode, we use a neural model to generate the sparse vector, just like we do during ingestion. Each query requires one model inference call. After we obtain the sparse vector, we transform it into Lucene disjunctive feature queries (`FeatureQuery(termA) OR FeatureQuery(termB) OR FeatureQuery(termC) ...`). Each `FeatureQuery` looks through the corresponding token's posting list. **Similarly to BM25, the search time complexity depends on the number of hit posting lists and the lists' lengths.** However, for neural sparse search, posting lists are longer (because of document expansion) and the bi-encoder needs to examine more posting lists (because of query expansion). Assuming all documents have a similar term frequency distribution, the length of posting lists grows proportionally to the number of documents. This means **the effort required to search the inverted index increases linearly with the number of documents in the dataset**.
```

Comment:
```suggestion
- For the **bi-encoder** mode, we use a neural model to generate the sparse vector, just like we do during ingestion. Each query requires one model inference call. After we obtain the sparse vector, we transform it into Lucene disjunctive feature queries (`FeatureQuery(termA) OR FeatureQuery(termB) OR FeatureQuery(termC) ...`). Each `FeatureQuery` looks through the corresponding token's posting list. **Similarly to BM25, the search time complexity depends on the number of hit posting lists and the lists' lengths.** However, for neural sparse search, posting lists are longer (because of document expansion) and the bi-encoder needs to examine more posting lists (because of query expansion). Assuming all documents have a similar term frequency distribution, the length of posting lists grows proportionally to the number of documents. This means that **the effort required to search the inverted index increases linearly with the number of documents in the dataset**.
```